lib/src/front_end/delimited_list_builder.dart

```diff
@@ -163,24 +102,40 @@ class DelimitedListBuilder {
     _rightBracket = _visitor.writer.pop();
   }
 
+  /// Adds [piece] to the built list.
+  ///
+  /// Use this when the piece is composed of more than one [AstNode] or [Token]
+  /// and [visit()] can't be used. When calling this, make sure to call
+  /// [addCommentsBefore()] for the first token in the [piece].
+  ///
+  /// Assumes there is no comma after this piece.
+  void add(Piece piece) {
+    _elements.add(ListElement(piece));
+    _commentsBeforeComma = CommentSequence.empty;
+  }
+
+  /// Writes any comments appearing before [token] to the list.
+  void addCommentsBefore(Token token) {
+    // Handle comments between the preceding element and this one.
+    var commentsBeforeElement = _visitor.takeCommentsBefore(token);
+    _addComments(commentsBeforeElement, hasElementAfter: true);
+  }
+
   /// Adds [element] to the built list.
   ///
   /// Includes any comments that appear before element. Also includes the
   /// subsequent comma, if any, and any comments that precede the comma.
-  void add(AstNode element) {
+  void visit(AstNode element) {
     // Handle comments between the preceding element and this one.
```

Comment:
This comment feels redundant against the clearly named method on the next line, and against the copied one inside the body of that method. We're in "repeat what the code does" territory with these comments (arguably with both copies).

Consider taking a pass find comments that repeat the code - noise like this makes the useful signal harder to find.

https://testing.googleblog.com/2017/07/code-health-to-comment-or-not-to-comment.html